clients/drcachesim/drpt2trace/pt2ir.cpp

```diff
@@ -0,0 +1,391 @@
+            if (status == -pte_eos)
+                break;
+            dx_decoding_error(status, "sync error", insn.ip);
+            return PT2IR_CONV_ERROR_SYNC_PACKET;
+        }
+
+        /* Decode the raw trace data surround by PSB. */
+        for (;;) {
+            int nextstatus = status;
+            int errcode = 0;
+            struct pt_image *image;
+
+            /* Before starting to decode instructions, we need to handle the event before
+             * the instruction trace. For example, if a mmap2 event happens, we need to
+             * switch the cached image.
+             */
+            while ((nextstatus & pts_event_pending) != 0) {
+                struct pt_event event;
+
+                nextstatus = pt_insn_event(pt_instr_decoder_, &event, sizeof(event));
+                if (nextstatus < 0) {
+                    errcode = nextstatus;
+                    dx_decoding_error(errcode, "get pending event error", insn.ip);
+                    return PT2IR_CONV_ERROR_GET_PENDING_EVENT;
+                }
+
+                /* Use a sideband session to check if pt_event is an image switch event.
+                 * If so, change the image in 'pt_instr_decoder_' to the target image.
+                 */
+                image = nullptr;
+                errcode =
+                    pt_sb_event(pt_sb_session_, &image, &event, sizeof(event), stdout, 0);
+                if (errcode < 0) {
+                    dx_decoding_error(errcode, "handle sideband event error", insn.ip);
+                    return PT2IR_CONV_ERROR_HANDLE_SIDEBAND_EVENT;
+                }
+
+                /* If it is not an image switch event, the PT instruction decoder
+                 * will not switch their cached image.
+                 */
+                if (image == nullptr)
+                    continue;
+
+                errcode = pt_insn_set_image(pt_instr_decoder_, image);
+                if (errcode < 0) {
+                    dx_decoding_error(errcode, "set image error", insn.ip);
+                    return PT2IR_CONV_ERROR_SET_IMAGE;
+                }
+            }
+            if ((nextstatus & pts_eos) != 0)
+                break;
+
+            /* Decode PT raw trace to pt_insn. */
+            status = pt_insn_next(pt_instr_decoder_, &insn, sizeof(insn));
+            if (status < 0) {
+                dx_decoding_error(status, "get next instruction error", insn.ip);
+                return PT2IR_CONV_ERROR_DECODE_NEXT_INSTR;
+            }
+
+            pt_insn_list_.push_back(insn);
+
+            /* TODO i#5505: Use drdecode to decode insn(pt_insn) to instr_t. */
+        }
+    }
+    return PT2IR_CONV_SUCCESS;
+}
+
+uint64_t
+pt2ir_t::get_instr_count()
+{
+    return pt_insn_list_.size();
+}
+
+void
+pt2ir_t::print_instrs_to_stdout()
+{
+    void *dcontext = GLOBAL_DCONTEXT;
+    for (auto pt_instr : pt_insn_list_) {
+        disassemble_with_info(dcontext, pt_instr.raw, STDOUT, false, false);
```

Comment:
It is nice to give bool or other less clear param names in comments like this:
```C
/*show_pc=*/false, /*show_bytes=*/false
```

Reply:
Done.